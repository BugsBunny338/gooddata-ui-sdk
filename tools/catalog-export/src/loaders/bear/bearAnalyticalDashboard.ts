// (C) 2021 GoodData Corporation
import gooddata from "@gooddata/api-client-bear";
import { ObjectMeta } from "../../base/types";

/**
 * Loads information about analytical dashboards defined in the project. Only descriptive information about
 * each analytical dashboard is returned.
 *
 * @param projectId - project to get analytical dashboards from
 * @return array of analytical dashboard meta
 */
export async function loadAnalyticalDashboard(projectId: string): Promise<ObjectMeta[]> {
    const analyticalDashboards = await gooddata.md.getAnalyticalDashboards(projectId);

    return analyticalDashboards.map((dashboard: any) => {
        return {
            identifier: dashboard.identifier,
            tags: dashboard.tags,
            title: dashboard.title,
        };
    });
}
