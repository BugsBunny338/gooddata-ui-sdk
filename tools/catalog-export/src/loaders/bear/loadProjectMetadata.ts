// (C) 2007-2019 GoodData Corporation
import ora from "ora";
import { logError } from "../../cli/loggers";
import { CatalogExportError, ProjectMetadata } from "../../base/types";
import { loadCatalog } from "./catalog";
import { loadDateDataSets } from "./dateDataSets";
import { loadInsights } from "./visualization";

/**
 * Loads all project metadata that can be used for exporting into catalog.
 *
 * @param projectId - project identifier
 * @throws CatalogExportError
 */
export async function loadProjectMetadata(projectId: string): Promise<ProjectMetadata> {
    const spinner = ora();

    try {
        spinner.start("Loading catalog of attributes and metrics…");
        const catalog = await loadCatalog(projectId);
        spinner.succeed("Catalog loaded");

        spinner.start("Loading date data sets…");
        const dateDataSets = await loadDateDataSets(projectId);
        spinner.succeed("Date data sets loaded");

        spinner.start("Loading insights…");
        const insights = await loadInsights(projectId);
        spinner.succeed("Insights loaded");

        return {
            projectId,
            catalog,
            dateDataSets,
            insights,
        };
    } catch (err) {
        spinner.fail();
        const more = err.response ? err.response.url : "";
        logError(`Exception: ${err.message} ${more}\n${err.stack}`);

        throw new CatalogExportError("A fatal error has occurred while loading project metadata", 1);
    }
}
