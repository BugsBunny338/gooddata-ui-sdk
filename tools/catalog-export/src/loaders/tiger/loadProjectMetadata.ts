// (C) 2007-2020 GoodData Corporation
import { CatalogExportError, DateDataSet, ProjectMetadata } from "../../base/types";
import { ITigerClient } from "@gooddata/gd-tiger-client";
import ora from "ora";
import { logError } from "../../cli/loggers";
import { loadCatalog } from "./catalog";
import { loadInsights } from "./insights";

// @ts-ignore
async function loadDateDataSets(_projectId: string, tigerClient: ITigerClient): Promise<DateDataSet[]> {
    return [];
}

export async function loadProjectMetadata(
    projectId: string,
    tigerClient: ITigerClient,
): Promise<ProjectMetadata> {
    const spinner = ora();

    try {
        spinner.start("Loading catalog of attributes and metrics…");
        const catalog = await loadCatalog(projectId, tigerClient);
        spinner.succeed("Catalog loaded");

        spinner.start("Loading date data sets…");
        const dateDataSets = await loadDateDataSets(projectId, tigerClient);
        spinner.succeed("Date data sets loaded");

        spinner.start("Loading insights…");
        const insights = await loadInsights(projectId, tigerClient);
        spinner.succeed("Insights loaded");

        return {
            projectId,
            catalog,
            dateDataSets,
            insights,
        };
    } catch (err) {
        spinner.fail();
        const more = err.response ? err.response.url : "";
        logError(`Exception: ${err.message} ${more}\n${err.stack}`);

        throw new CatalogExportError("A fatal error has occurred while loading project metadata", 1);
    }
}
