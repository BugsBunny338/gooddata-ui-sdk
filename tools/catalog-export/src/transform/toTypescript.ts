// (C) 2007-2019 GoodData Corporation
import { Attribute, DateDataSet, DisplayForm, Fact, Metric, ProjectMetadata } from "../base/types";
import {
    ImportDeclarationStructure,
    OptionalKind,
    Project,
    SourceFile,
    VariableDeclarationKind,
    VariableStatementStructure,
} from "ts-morph";
import { createUniqueVariableName, TakenNamesMap } from "./titles";
import { flatMap } from "lodash";

export type TypescriptOutput = {
    project: Project;
    sourceFile: SourceFile;
};

//
// Internal functions to build variable names for various
//

let GlobalNameScope: TakenNamesMap = {};

type NamingStrategy = (title: string, scope: TakenNamesMap) => string;

type AttributeNaming = {
    attribute: NamingStrategy;
    displayForm: NamingStrategy;
};

const DefaultNaming: AttributeNaming = {
    attribute: uniqueVariable,
    displayForm: uniqueVariable,
};

const DateDataSetNaming: AttributeNaming = {
    attribute: dateAttributeSwitcharoo,
    displayForm: dateDisplayFormStrip,
};

/**
 * This is a wrapper on top of createUniqueVariableName() which mutates the input map of used
 * variable names.
 *
 * @param title - md object title
 * @param nameScope - scope containing already taken variable names, defaults to global scope
 */
function uniqueVariable(title: string, nameScope: TakenNamesMap = GlobalNameScope): string {
    const variableName = createUniqueVariableName(title, nameScope);
    nameScope[variableName] = true;

    return variableName;
}

/**
 * This is a wrapper on top of uniqueVariable. It is useful when naming date data set attributes. They have
 * a convention where date data set name is in parenthesis at the end of the attr/df title. This function
 * takes the ds name and moves it at the beginning of the title. That way all variables for same date data set
 * start with the same prefix.
 *
 * @param title - title to play with
 * @param nameScope - scope containing already taken variable names
 */
function dateAttributeSwitcharoo(title: string, nameScope: TakenNamesMap = GlobalNameScope): string {
    const datasetStart = title.lastIndexOf("(");
    const switchedTitle = `${title.substr(datasetStart)} ${title.substr(0, datasetStart)}`;

    return uniqueVariable(switchedTitle, nameScope);
}

function dateDisplayFormStrip(title: string, nameScope: TakenNamesMap = GlobalNameScope): string {
    const metaStart = title.indexOf("(");

    return uniqueVariable(title.substr(0, metaStart), nameScope);
}

//
// Transformation to ts-morph structures
//

function initialize(outputFile: string): TypescriptOutput {
    const project = new Project({});
    const sourceFile = project.createSourceFile(outputFile);

    return {
        project,
        sourceFile,
    };
}

function generateSdkModelImports(): OptionalKind<ImportDeclarationStructure> {
    return {
        moduleSpecifier: "@gooddata/sdk-model",
        namedImports: ["newAttribute", "newMeasure", "IAttribute", "IMeasure", "IMeasureDefinition"],
    };
}

function generateAttributeDisplayForm(
    displayForm: DisplayForm,
    attributeVariableName: string,
    nameScope: TakenNamesMap,
    naming: AttributeNaming,
): string {
    const { meta } = displayForm;
    const dfVariableName = naming.displayForm(meta.title, nameScope);
    let variableName = attributeVariableName === dfVariableName ? "Default" : dfVariableName;

    if (variableName.startsWith(attributeVariableName)) {
        variableName = variableName.substr(attributeVariableName.length);
    }

    return `/** \n* Display Form Title: ${meta.title}  \n* Display Form ID: ${meta.identifier}\n*/\n${variableName}: newAttribute('${meta.identifier}')`;
}

function generateAttribute(
    attribute: Attribute,
    naming: AttributeNaming = DefaultNaming,
): OptionalKind<VariableStatementStructure> {
    const { meta } = attribute.attribute;
    const variableName = naming.attribute(meta.title, GlobalNameScope);
    const { displayForms } = attribute.attribute.content;

    if (displayForms.length === 1) {
        /*
         * If there is a single DF for the attribute, then have const AttrName = newAttribute(...)
         */
        const displayForm = displayForms[0];

        return {
            declarationKind: VariableDeclarationKind.Const,
            isExported: true,
            docs: [`Attribute Title: ${meta.title}\nDisplay Form ID: ${meta.identifier}`],
            declarations: [
                {
                    name: variableName,
                    type: "IAttribute",
                    initializer: `newAttribute('${displayForm.meta.identifier}')`,
                },
            ],
        };
    } else {
        /*
         * If there are multiple DFs, have mapping of const AttrName = { DfName: newAttribute(), OtherDfName: newAttribute()}
         */
        const localNameScope: TakenNamesMap = {};
        const displayFormInits: string[] = attribute.attribute.content.displayForms.map(df =>
            generateAttributeDisplayForm(df, variableName, localNameScope, naming),
        );

        return {
            declarationKind: VariableDeclarationKind.Const,
            isExported: true,
            declarations: [
                {
                    name: variableName,
                    type: "{[df: string]: IAttribute}",
                    initializer: `{ ${displayFormInits.join(",")} }`,
                },
            ],
        };
    }
}

function generateAttributes(
    projectMeta: ProjectMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    return projectMeta.catalog.attributes.map(a => generateAttribute(a));
}

function generateMeasureFromMetric(metric: Metric): OptionalKind<VariableStatementStructure> {
    const { meta } = metric.metric;
    const variableName = uniqueVariable(meta.title);

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        docs: [`Metric Title: ${meta.title}\nMetric ID: ${meta.identifier}\nMetric Type: MAQL Metric`],
        declarations: [
            {
                name: variableName,
                type: "IMeasure<IMeasureDefinition>",
                initializer: `newMeasure('${metric.metric.meta.identifier}')`,
            },
        ],
    };
}

const aggregations = ["sum", "count", "avg", "min", "max", "median", "runsum"];

function generateFactAggregations(fact: Fact): string[] {
    const { meta } = fact.fact;

    return aggregations.map(aggregation => {
        const jsDoc = `/** \n* Fact Title: ${meta.title}  \n* Fact ID: ${meta.identifier}\n * Fact Aggregation: ${aggregation}\n*/`;
        const name = aggregation.charAt(0).toUpperCase() + aggregation.substr(1);

        return `${jsDoc}\n${name}: newMeasure('${meta.identifier}', m => m.aggregation('${aggregation}'))`;
    });
}

function generateMeasuresFromFacts(fact: Fact): OptionalKind<VariableStatementStructure> {
    const { meta } = fact.fact;
    const variableName = uniqueVariable(meta.title);

    const aggregationInits: string[] = generateFactAggregations(fact);

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        docs: [`Fact Title: ${meta.title}\nFact ID: ${meta.identifier}`],
        declarations: [
            {
                name: variableName,
                type: "{[measure: string]: IMeasure<IMeasureDefinition>}",
                initializer: `{ ${aggregationInits.join(",")} } `,
            },
        ],
    };
}

function generateMeasures(
    projectMeta: ProjectMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    const fromMetrics = projectMeta.catalog.metrics.map(generateMeasureFromMetric);
    const fromFacts = projectMeta.catalog.facts.map(generateMeasuresFromFacts);

    return fromMetrics.concat(fromFacts);
}

function generateDateDataSet(dd: DateDataSet): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    const { content } = dd.dateDataSet;

    return content.attributes.map(a => generateAttribute(a, DateDataSetNaming));
}

function generateDateDataSets(
    projectMeta: ProjectMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    return flatMap(projectMeta.dateDataSets.map(generateDateDataSet));
}

/**
 * Transforms project metadata into model definitions in TypeScript. The resulting TS source file will contain
 * constant declarations for the various objects in the metadata:
 *
 * - Attributes with single display form are transformed to constants initialized with respective newAttribute()
 * - Attributes with multiple display forms are transformed to constants initialized to map of df name => newAttribute()
 * - Metrics (MAQL) will be transformed to constants initialized with respective newMeasure()
 * - Metrics from facts will be transformed to constants initialized to map of aggregation => newMeasure()
 * - Date data set attributes will be transformed using the same logic as normal attributes, albeit with slightly
 *   modified variable naming strategy
 *
 * @param projectMeta - project metadata to transform to typescript
 * @param outputFile - output typescript file
 * @return return of the transformation process, new file is not saved at this point
 */
export async function transformToTypescript(
    projectMeta: ProjectMetadata,
    outputFile: string,
): Promise<TypescriptOutput> {
    GlobalNameScope = {};

    const output = initialize(outputFile);
    const { sourceFile } = output;

    sourceFile.addImportDeclaration(generateSdkModelImports());
    sourceFile.addVariableStatements(generateAttributes(projectMeta));
    sourceFile.addVariableStatements(generateMeasures(projectMeta));
    sourceFile.addVariableStatements(generateDateDataSets(projectMeta));

    return output;
}
