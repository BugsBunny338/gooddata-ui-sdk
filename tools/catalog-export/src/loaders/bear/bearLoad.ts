// (C) 2007-2021 GoodData Corporation
import ora from "ora";
import { logError } from "../../cli/loggers";
import { CatalogExportError, ProjectMetadata } from "../../base/types";
import { loadCatalog } from "./bearCatalog";
import { loadDateDataSets } from "./bearDateDatasets";
import { loadInsights } from "./bearInsights";
import { loadAnalyticalDashboard } from "./bearAnalyticalDashboard";

/**
 * Loads all project metadata that can be used for exporting into catalog.
 *
 * @param projectId - project identifier
 * @throws CatalogExportError
 */
export async function bearLoad(projectId: string): Promise<ProjectMetadata> {
    const spinner = ora();

    try {
        spinner.start("Loading catalog of attributes and metrics…");
        const catalog = await loadCatalog(projectId);
        spinner.succeed("Catalog loaded");

        spinner.start("Loading date data sets…");
        const dateDataSets = await loadDateDataSets(projectId);
        spinner.succeed("Date data sets loaded");

        spinner.start("Loading insights…");
        const insights = await loadInsights(projectId);
        spinner.succeed("Insights loaded");

        spinner.start("Loading analytical dashboards…");
        const analyticalDashboards = await loadAnalyticalDashboard(projectId);
        spinner.succeed("Analytical dashboards loaded");
        return {
            projectId,
            catalog,
            dateDataSets,
            insights,
            analyticalDashboards,
        };
    } catch (err) {
        spinner.fail();
        const more = err.response ? err.response.url : "";
        logError(`Exception: ${err.message} ${more}\n${err.stack}`);

        throw new CatalogExportError("A fatal error has occurred while loading project metadata", 1);
    }
}
