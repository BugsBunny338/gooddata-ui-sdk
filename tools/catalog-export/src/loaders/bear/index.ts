// (C) 2007-2020 GoodData Corporation

import { CatalogExportConfig, CatalogExportError, ProjectMetadata } from "../../base/types";
import { DEFAULT_HOSTNAME } from "../../base/constants";
import * as pkg from "../../../package.json";
import ora from "ora";
import { log, logError } from "../../cli/loggers";
import { promptPassword, promptProjectId, promptUsername } from "../../cli/prompts";
import { clearLine } from "../../cli/clear";
import gooddata from "@gooddata/gd-bear-client";
import { loadProjectMetadata } from "./loadProjectMetadata";

/**
 * Given the export config, ask for any missing information and then load project metadata from
 * a bear project.
 *
 * @param config - tool configuration, may be missing username, password and project id - in that case code
 *  will promp
 *
 * @returns loaded project metadata
 *
 * @throws CatalogExportError upon any error.
 */
export async function loadProjectMetadataFromBackend(config: CatalogExportConfig): Promise<ProjectMetadata> {
    const { projectName, hostname } = config;
    let { projectId, username, password } = config;

    gooddata.config.setCustomDomain(hostname || DEFAULT_HOSTNAME);
    gooddata.config.setJsPackage(pkg.name, pkg.version);

    const logInSpinner = ora();
    try {
        if (username) {
            log("Username", username);
        } else {
            username = await promptUsername();
        }

        password = password || (await promptPassword());

        logInSpinner.start("Logging in...");
        await gooddata.user.login(username, password);
        logInSpinner.stop();
        clearLine();
    } catch (err) {
        logInSpinner.fail();
        clearLine();

        if (err.message && err.message.search(/.*(certificate|self-signed).*/) > -1) {
            logError(
                "Server does not have valid certificate. The login has failed. " +
                    "If you trust the server, you can use the --accept-untrusted-ssl option " +
                    "to turn off certificate validation.",
            );
        }

        throw new CatalogExportError(`Unable to log in to platform. The error was: ${err}`, 1);
    }

    const projectSpinner = ora();
    try {
        if (projectName && !projectId) {
            log("Project Name", projectName);
            projectSpinner.start("Loading project");
            const metadataResponse = await gooddata.xhr.get("/gdc/md");
            const metadata = metadataResponse.getData();
            projectSpinner.stop();
            const projectMetadata = metadata.about
                ? metadata.about.links.find((link: any) => {
                      return link.title === projectName;
                  })
                : null;
            if (projectMetadata) {
                projectId = projectMetadata.identifier;
            } else {
                logError(`Could not find a project with name '${projectName}'`);
            }
        }
        if (projectId) {
            log("Project ID", projectId);
        } else {
            projectId = await promptProjectId();
        }

        return loadProjectMetadata(projectId);
    } catch (err) {
        projectSpinner.stop();

        throw new CatalogExportError(
            `Unable to obtain project metadata from platform. The error was: ${err}`,
            1,
        );
    }
}
