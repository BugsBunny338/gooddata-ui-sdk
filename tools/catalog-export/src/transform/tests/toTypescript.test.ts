// (C) 2007-2020 GoodData Corporation
import { loadProjectMetadata } from "../../loaders/bear/loadProjectMetadata";
import { transformToTypescript } from "../toTypescript";

jest.mock("@gooddata/gd-bear-client");

describe("transformToCatalog", () => {
    const projectMeta = loadProjectMetadata("test");

    it("creates new catalog", async () => {
        const transformResult = transformToTypescript(await projectMeta, "testOutput.ts", false);

        /*
         * NOTE: source.getFullText() should not be used here as it includes the leading comments that
         * contains timestamp describing when was the file generated.
         */
        expect(transformResult.sourceFile.getText()).toMatchSnapshot();
    });
});
