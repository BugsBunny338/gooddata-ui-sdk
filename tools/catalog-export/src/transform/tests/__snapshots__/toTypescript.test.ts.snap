// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformToCatalog creates new catalog 1`] = `
"import { newAttribute, newMeasure, IAttribute, IMeasure, IMeasureDefinition, idRef } from \\"@gooddata/sdk-model\\";

/**
 * Attribute Title: Channel
 * Display Form ID: attr.orders.channel
 */
export const Channel: IAttribute = newAttribute('label.orders.channel');
export const EmployeeName = { /** 
    * Display Form Title: Employee Name  
    * Display Form ID: label.employee.employeename
    */
    Default: newAttribute('label.employee.employeename'),/** 
    * Display Form Title: Employee URL  
    * Display Form ID: label.employee.employeename.employeeurl
    */
    EmployeeURL: newAttribute('label.employee.employeename.employeeurl') };
/**
 * Metric Title: _Filter Last 4 Quarter
 * Metric ID: aaIHiWZjfWNA
 * Metric Type: MAQL Metric
 */
export const FilterLast4Quarter: IMeasure<IMeasureDefinition> = newMeasure(idRef('aaIHiWZjfWNA', \\"measure\\"));
/**
 * Metric Title: _Filter Quarter
 * Metric ID: aaiF4bffe4yn
 * Metric Type: MAQL Metric
 */
export const FilterQuarter: IMeasure<IMeasureDefinition> = newMeasure(idRef('aaiF4bffe4yn', \\"measure\\"));
/**
 * Fact Title: $ Menu Item Sales
 * Fact ID: fact.salesdetailfact.menuitemsales
 */
export const $MenuItemSales = { /** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: sum
    */
    Sum: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('sum')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: count
    */
    Count: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('count')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: avg
    */
    Avg: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('avg')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: min
    */
    Min: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('min')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: max
    */
    Max: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('max')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: median
    */
    Median: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('median')),/** 
    * Fact Title: $ Menu Item Sales  
    * Fact ID: fact.salesdetailfact.menuitemsales
     * Fact Aggregation: runsum
    */
    Runsum: newMeasure(idRef('fact.salesdetailfact.menuitemsales', \\"fact\\"), m => m.aggregation('runsum')) } ;
/**
 * Fact Title: Cost
 * Fact ID: fact.restaurantcostsfact.cost
 */
export const Cost = { /** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: sum
    */
    Sum: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('sum')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: count
    */
    Count: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('count')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: avg
    */
    Avg: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('avg')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: min
    */
    Min: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('min')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: max
    */
    Max: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('max')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: median
    */
    Median: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('median')),/** 
    * Fact Title: Cost  
    * Fact ID: fact.restaurantcostsfact.cost
     * Fact Aggregation: runsum
    */
    Runsum: newMeasure(idRef('fact.restaurantcostsfact.cost', \\"fact\\"), m => m.aggregation('runsum')) } ;
/**
 * Attribute Title: Year (Timeline)
 * Display Form ID: timeline.year
 */
export const TimelineYear: IAttribute = newAttribute('timeline.aag81lMifn6q');
/**
 * Attribute Title: Quarter (Timeline)
 * Display Form ID: timeline.quarter.in.year
 */
export const TimelineQuarter: IAttribute = newAttribute('timeline.aam81lMifn6q');
export const TimelineWeekSunSatYear = { /** 
    * Display Form Title: Week #/Year (W1/2010) (Timeline)  
    * Display Form ID: timeline.aaA81lMifn6q
    */
    WeekNrYear: newAttribute('timeline.aaA81lMifn6q'),/** 
    * Display Form Title: Week Starting (Timeline)  
    * Display Form ID: timeline.aaw81lMifn6q
    */
    WeekStarting: newAttribute('timeline.aaw81lMifn6q'),/** 
    * Display Form Title: From - To (Timeline)  
    * Display Form ID: timeline.aau81lMifn6q
    */
    FromTo: newAttribute('timeline.aau81lMifn6q'),/** 
    * Display Form Title: Week #/Year (Cont.) (Timeline)  
    * Display Form ID: timeline.aay81lMifn6q
    */
    WeekNrYear_1: newAttribute('timeline.aay81lMifn6q'),/** 
    * Display Form Title: Wk/Qtr/Year (Cont.) (Timeline)  
    * Display Form ID: timeline.aaC81lMifn6q
    */
    WkQtrYear: newAttribute('timeline.aaC81lMifn6q'),/** 
    * Display Form Title: Wk/Qtr/Year (Timeline)  
    * Display Form ID: timeline.aas81lMifn6q
    */
    WkQtrYear_1: newAttribute('timeline.aas81lMifn6q') };
/**
 * Attribute Title: Week (Sun-Sat) (Timeline)
 * Display Form ID: timeline.week.in.year
 */
export const TimelineWeekSunSat: IAttribute = newAttribute('timeline.aaI81lMifn6q');
/**
 * Attribute Title: Week (Sun-Sat) of Qtr (Timeline)
 * Display Form ID: timeline.week.in.quarter
 */
export const TimelineWeekSunSatOfQtr: IAttribute = newAttribute('timeline.aaO81lMifn6q');
export const TimelineWeekMonSunYear = { /** 
    * Display Form Title: Week #/Year (W1/2010) (Timeline)  
    * Display Form ID: timeline.aa281lMifn6q
    */
    WeekNrYear: newAttribute('timeline.aa281lMifn6q'),/** 
    * Display Form Title: Week Starting (Timeline)  
    * Display Form ID: timeline.aaY81lMifn6q
    */
    WeekStarting: newAttribute('timeline.aaY81lMifn6q'),/** 
    * Display Form Title: From - To (Timeline)  
    * Display Form ID: timeline.aaW81lMifn6q
    */
    FromTo: newAttribute('timeline.aaW81lMifn6q') };
/**
 * Attribute Title: Week (Mon-Sun) (Timeline)
 * Display Form ID: timeline.euweek.in.year
 */
export const TimelineWeekMonSun: IAttribute = newAttribute('timeline.aba81lMifn6q');
/**
 * Attribute Title: Week (Mon-Sun) of Qtr (Timeline)
 * Display Form ID: timeline.euweek.in.quarter
 */
export const TimelineWeekMonSunOfQtr: IAttribute = newAttribute('timeline.abg81lMifn6q');
export const TimelineMonth = { /** 
    * Display Form Title: Short (Jan) (Timeline)  
    * Display Form ID: timeline.abm81lMifn6q
    */
    Short: newAttribute('timeline.abm81lMifn6q'),/** 
    * Display Form Title: Long (January) (Timeline)  
    * Display Form ID: timeline.abs81lMifn6q
    */
    Long: newAttribute('timeline.abs81lMifn6q'),/** 
    * Display Form Title: Number (M1) (Timeline)  
    * Display Form ID: timeline.abq81lMifn6q
    */
    Number: newAttribute('timeline.abq81lMifn6q'),/** 
    * Display Form Title: M/Q (M1/Q1) (Timeline)  
    * Display Form ID: timeline.abo81lMifn6q
    */
    MQ: newAttribute('timeline.abo81lMifn6q') };
/**
 * Attribute Title: Month of Quarter (Timeline)
 * Display Form ID: timeline.month.in.quarter
 */
export const TimelineMonthOfQuarter: IAttribute = newAttribute('timeline.aby81lMifn6q');
/**
 * Attribute Title: Day of Year (Timeline)
 * Display Form ID: timeline.day.in.year
 */
export const TimelineDayOfYear: IAttribute = newAttribute('timeline.abE81lMifn6q');
export const TimelineDayOfWeekSunSat = { /** 
    * Display Form Title: Short (Sun) (Timeline)  
    * Display Form ID: timeline.abK81lMifn6q
    */
    Short: newAttribute('timeline.abK81lMifn6q'),/** 
    * Display Form Title: Long (Sunday) (Timeline)  
    * Display Form ID: timeline.abO81lMifn6q
    */
    Long: newAttribute('timeline.abO81lMifn6q'),/** 
    * Display Form Title: Number (1=Sunday) (Timeline)  
    * Display Form ID: timeline.abM81lMifn6q
    */
    Number: newAttribute('timeline.abM81lMifn6q') };
export const TimelineDayOfWeekMonSun = { /** 
    * Display Form Title: Short (Mon) (Timeline)  
    * Display Form ID: timeline.abU81lMifn6q
    */
    Short: newAttribute('timeline.abU81lMifn6q'),/** 
    * Display Form Title: Long (Monday) (Timeline)  
    * Display Form ID: timeline.abY81lMifn6q
    */
    Long: newAttribute('timeline.abY81lMifn6q'),/** 
    * Display Form Title: Number (1=Monday) (Timeline)  
    * Display Form ID: timeline.abW81lMifn6q
    */
    Number: newAttribute('timeline.abW81lMifn6q') };
/**
 * Attribute Title: Day of Quarter (Timeline)
 * Display Form ID: timeline.day.in.quarter
 */
export const TimelineDayOfQuarter: IAttribute = newAttribute('timeline.ab481lMifn6q');
/**
 * Attribute Title: Day of Month (Timeline)
 * Display Form ID: timeline.day.in.month
 */
export const TimelineDayOfMonth: IAttribute = newAttribute('timeline.aca81lMifn6q');
/**
 * Attribute Title: Quarter/Year (Timeline)
 * Display Form ID: timeline.quarter
 */
export const TimelineQuarterYear: IAttribute = newAttribute('timeline.aci81lMifn6q');
export const TimelineMonthYear = { /** 
    * Display Form Title: Short (Jan 2010) (Timeline)  
    * Display Form ID: timeline.act81lMifn6q
    */
    Short: newAttribute('timeline.act81lMifn6q'),/** 
    * Display Form Title: Long (January 2010) (Timeline)  
    * Display Form ID: timeline.acx81lMifn6q
    */
    Long: newAttribute('timeline.acx81lMifn6q'),/** 
    * Display Form Title: Number (1/2010) (Timeline)  
    * Display Form ID: timeline.acv81lMifn6q
    */
    Number: newAttribute('timeline.acv81lMifn6q') };
export const TimelineDate = { /** 
    * Display Form Title: mm/dd/yyyy (Timeline)  
    * Display Form ID: timeline.date.mmddyyyy
    */
    MmDdYyyy: newAttribute('timeline.date.mmddyyyy'),/** 
    * Display Form Title: yyyy-mm-dd (Timeline)  
    * Display Form ID: timeline.date.yyyymmdd
    */
    YyyyMmDd: newAttribute('timeline.date.yyyymmdd'),/** 
    * Display Form Title: m/d/yy (no leading zeroes) (Timeline)  
    * Display Form ID: timeline.date.mdyy
    */
    MDYy: newAttribute('timeline.date.mdyy'),/** 
    * Display Form Title: Long (Mon, Jan 1, 2010) (Timeline)  
    * Display Form ID: timeline.date.long
    */
    Long: newAttribute('timeline.date.long'),/** 
    * Display Form Title: dd/mm/yyyy (Timeline)  
    * Display Form ID: timeline.date.ddmmyyyy
    */
    DdMmYyyy: newAttribute('timeline.date.ddmmyyyy'),/** 
    * Display Form Title: dd-mm-yyyy (Timeline)  
    * Display Form ID: timeline.date.eddmmyyyy
    */
    DdMmYyyy_1: newAttribute('timeline.date.eddmmyyyy') };
export const Insights = { /** 
    * Insight Title: Labor Costs vs Scheduled Costs  
    * Insight ID: abhJpedgcfU2
    */
    LaborCostsVsScheduledCosts: 'abhJpedgcfU2',/** 
    * Insight Title: Total Sales by City  
    * Insight ID: acPWDTL2bJeX
    */
    TotalSalesByCity: 'acPWDTL2bJeX',/** 
    * Insight Title: $ Total Sales by Location  
    * Insight ID: aaBN5UG3dXu4
    */
    $TotalSalesByLocation: 'aaBN5UG3dXu4' };
"
`;
